README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# UseriesGAwigl

<!-- badges: start -->
<!-- badges: end -->

UseriesGAwigl calculates weathering ages from uranium-series activity
ratios in weathering profiles, using a genetic algorithm for
optimisation

## Web application

The package includes a web application, online at
<https://anthony-dosseto.shinyapps.io/U_series_gain_loss_optimisation/>,
where you can upload your CSV file, set the model parameters, run the
model, and view the results. Choose this option is you are not familiar
with R.

## Installation

You can install the development version of UseriesGAwigl from
[GitHub](https://github.com/tonydoss/UseriesGAwigl) with:

``` r
if(!require("remotes")) install.packages("remotes")
remotes::install_github("tonydoss/UseriesGAwigl")
```

## Example


NEWS.md
# UseriesGAwigl 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(UseriesGAwigl)

test_check("UseriesGAwigl")


tests/testthat/test-U_series_gain_loss_function.R
result <- U_series_gain_loss_function (read.csv(system.file("extdata", "data.csv", package = "UseriesGAwigl")), nbit = 100,
                                       logT_min = 1, logT_max = 7, Tmult_min = 0.5,
                                       logk238_min = -7, logk238_max = -5, k48_min = 0.1, k48_max = 10,
                                       logf238_k238_min = -20, logf238_k238_max = 3, f48_min = 0.1, f48_max = 10)

test_that("Model returns output", {
  expect_output(str(result), "List of 7")
})


R/U_series_gain_loss_function.R
globalVariables(c("R234U_238U", "R230Th_238U", "Depth"))

#' U_series_gain_loss_function
#'
#' @param input_data Read a CSV file with the required data. For instance, if the CSV file is called "data.csv" and is in your working directory, have input_data = read.csv("data.csv")
#' @param nbit Number of iterations
#' @param logT_min Log10 of the minimum value allowed for the weathering age of the shallowest sample
#' @param logT_max Log10 of the maximum value allowed for the weathering age of the shallowest sample
#' @param Tmult_min Minimum value allowed for Tmult
#' @param logk238_min Log10 of the minimum value allowed for k238
#' @param logk238_max Log10 of the maximum value allowed for k238
#' @param k48_min Minimum value allowed for k234/k238
#' @param k48_max Maximum value allowed for k234/k238
#' @param logf238_k238_min Log10 of the minimum value allowed for f238/k238
#' @param logf238_k238_max Log10 of the maximum value allowed for f238/k238
#' @param f48_min Minimum value allowed for f234/f238
#' @param f48_max Maximum value allowed for f234/f238
#'
#' @return
#' @export
#'
#' @examples U_series_gain_loss_function (read.csv(system.file("extdata", "data.csv",
#' package = "UseriesGAwigl")),
#' logT_min = 1, logT_max = 7, Tmult_min = 0.5,
#' logk238_min = -7, logk238_max = -5,
#' k48_min = 0.1, k48_max = 10,
#' logf238_k238_min = -20, logf238_k238_max = 3,
#' f48_min = 0.1, f48_max = 10)
U_series_gain_loss_function <- function(
  input_data,
  nbit = 1000,
  logT_min = 1,
  logT_max = 7,
  Tmult_min = 0.5,
  logk238_min = -7,
  logk238_max = -5,
  k48_min = 1,
  k48_max = 2,
  logf238_k238_min = -8,
  logf238_k238_max = +0.5,
  f48_min = 0.1,
  f48_max = 10
){

  # # Example
  # U_series_gain_loss_function (read.csv("data.csv"),
  # logT_min = 1, logT_max = 7, Tmult_min = 0.5,
  # logk238_min = -7, logk238_max = -5, k48_min = 0.1, k48_max = 10,
  # logf238_k238_min = -20, logf238_k238_max = 3, f48_min = 0.1, f48_max = 10)

  # Import data -------------------------------------------------------------

  df <- input_data # this calls in our data we've stored in a csv format.
  df <- df[order(df$Depth..m.),] # sort data frame by increasing depth
  nb_sample <- nrow(df)-1 # nb of samples to solve - exclude deepest sample in solving, since used as initial conditions (so weathering age should be 0 a, which the model might have a hard time achieving)

  # Set variables for genetic algorithm -------------------------------------

  # nbrun <- 100 # the number of consecutive generations without any improvement in the best fitness value before the GA is stopped
  # elitism <- 10 # the number of best fitness individuals to survive at each generation. By default the top 5% individuals will survive at each iteration

  # Set inital range of values -----------------------------------------------------

  Tmult_max <- 1

  # range of values allowed for (234U/238U) at t = 0
  r48i_min <- df$R234U_238U[which(df$Depth..m.==max(df$Depth..m.))]-2*df$R48_2SE[which(df$Depth..m.==max(df$Depth..m.))]
  r48i_max <- df$R234U_238U[which(df$Depth..m.==max(df$Depth..m.))]+2*df$R48_2SE[which(df$Depth..m.==max(df$Depth..m.))]
  # range of values allowed for (230Th/238U) at t = 0
  r08i_min <- df$R230Th_238U[which(df$Depth..m.==max(df$Depth..m.))]-2*df$R08_2SE[which(df$Depth..m.==max(df$Depth..m.))]
  r08i_max <- df$R230Th_238U[which(df$Depth..m.==max(df$Depth..m.))]+2*df$R08_2SE[which(df$Depth..m.==max(df$Depth..m.))]

  min_val <- c(logk238_min, k48_min, logf238_k238_min,
               f48_min, r48i_min, r08i_min, logT_min, rep(Tmult_min,
                                                          nb_sample - 1))

  max_val <- c(logk238_max, k48_max, logf238_k238_max,
               f48_max, r48i_max, r08i_max, logT_max, rep(Tmult_max,
                                                          nb_sample - 1))

  # Constants ---------------------------------------------------------------


  # values for 230Th and 232Th are set
  k230 <- 1e-10
  k232 <- 1e-10
  f230 <- 1e-10
  k232 <- 1e-10

  # decay constants
  l238 <- 0.1551e-9
  l234 <- 2.826e-6
  l232 <- 4.948e-11
  l230 <- 9.158e-6

  # U concentration at t=0 (in ppm); value doesn't matter
  U_init <- 2

  # Initialise vectors ------------------------------------------------------


  # initialise vectors and matrices used in loops
  R48_target <- vector(mode="numeric", length=nb_sample)
  R08_target <- vector(mode="numeric", length=nb_sample)
  R48_err <- vector(mode="numeric", length=nb_sample)
  R08_err <- vector(mode="numeric", length=nb_sample)
  depth <- vector(mode="numeric", length=nb_sample)
  T <- vector(mode="numeric", length=nb_sample)
  err_T <- vector(mode="numeric", length=nb_sample)
  R48_calc <- vector(mode="numeric", length=nb_sample)
  R08_calc <- vector(mode="numeric", length=nb_sample)
  f <- vector(mode="numeric", length=1)


  # set target ratios -------------------------------------------------------


  for (j in 1:nb_sample){
    R48_target[j] <- df$R234U_238U[j]
    R08_target[j] <- df$R230Th_238U[j]
    depth[j] <- as.matrix(df$Depth..m.[j])
    R48_err[j] <- df$R48_2SE[j]
    R08_err[j] <- df$R08_2SE[j]
  }

  target_ratios <- rbind(depth, R48_target, R48_err, R08_target, R08_err)
  target_ratios <- as.data.frame(t(target_ratios))


  # function to minimise ----------------------------------------------------


  fn <- function(x) {
    k238 <- 10^as.numeric(x[1])
    k234 <- k238*as.numeric(x[2])
    f238 <- k238*10^as.numeric(x[3])
    f234 <- f238*as.numeric(x[4])
    r48i <- as.numeric(x[5])
    r08i <- as.numeric(x[6])

    for (j in 1:nb_sample){
      if (j == 1){
        T[j] <-  10^as.numeric(x[7])}
      else {
        T[j] <- T[j-1]*as.numeric(x[6+j])
      }
    }

    r04i <- r08i/r48i
    U8i <- U_init*1e-6/238*6.02e23
    U4i <- r48i*l238/l234*U8i
    Th0i <- r04i*l234/l230*U4i

    F238 <- f238*l238*U8i
    F234 <- f234*l234*U4i
    F230 <- f230*l230*Th0i

    a4 <- l234 + k234
    a8 <- l238 + k238
    a0 <- l230 + k230

    A <- l230*l234/((a0 - a8)*(a4 - a8)) - l230*l234*F238/(l238*U8i*a8*(a4 - a8)*(a0 - a8))
    B <- l230/(a0 - a4)*r48i - l230*l234/((a0 - a4)*(a4 - a8)) - l230*F234/(l238*U8i*a4*(a0 - a4)) - l230*l234*F238/(l238*U8i*a4*(a0 - a4)*(a4 - a8))
    C <- r08i - l230/(a0 - a4)*r48i + l230*l234/((a0 - a4)*(a4 - a8)) - l230*l234/((a0 - a8)*(a4 - a8)) - F230/(l238*U8i*a0) - l230*F234/(l238*U8i*a0*a4) + l230*F234/(l238*U8i*a4*(a0 - a4)) - l230*l234*F238/(l238*U8i*a0*a4*a8) + l230*l234*F238/(l238*U8i*a4*a8*(a0 - a4)) - l230*l234*F238/(l238*U8i*a8*(a0 - a4)*(a4 - a8)) + l230*l234*F238/(l238*U8i*a8*(a0 - a8)*(a4 - a8))

    for (j in 1:nb_sample){
      R48_calc[j] <- ((l234/(a4 - a8) - l234*F238/(l238*U8i*a8*(a4 - a8)))*exp(-a8*T[j]) + (r48i - l234/(a4 - a8) - l234*F238/(l238*U8i*a4*(a4 - a8)) - F234/(l238*U8i*a4))*exp(-a4*T[j]) + F234/(l238*U8i*a4) + l234*F238/(l238*U8i*a4*a8))/((1 - F238/(l238*U8i*a8))*exp(-a8*T[j]) + F238/(l238*U8i*a8))
      R08_calc[j] <- (A*exp(-a8*T[j]) + B*exp(-a4*T[j]) + C*exp(-a0*T[j]) + l230*l234*F238/(l238*U8i*a0*a4*a8) + l230*F234/(l238*U8i*a0*a4) + F230/(l238*U8i*a0))/((1 - F238/(l238*U8i*a8))*exp(-a8*T[j]) + F238/(l238*U8i*a8))
    }

    f <- sum((R48_calc - R48_target)^2 + (R08_calc - R08_target)^2)
  }


  # solve -------------------------------------------------------------------
  # if(!require(GA)){
  #   install.packages("GA")
  #   require(GA)}

  GA <- GA::ga(type = "real-valued", fitness = function(x) -fn(x),
           lower = min_val, upper = max_val,
           # elitism = elitism, run = nbrun,
           maxiter = nbit, seed = 123, monitor = F)

  solution <- GA@solution


  # produce ratios ----------------------------------------------------------

  k238 <- 10^as.numeric(solution[1])
  k234 <- k238*as.numeric(solution[2])
  f238 <- k238*10^as.numeric(solution[3])
  f234 <- f238*as.numeric(solution[4])
  r48i <- as.numeric(solution[5])
  r08i <- as.numeric(solution[6])

  for (j in 1:nb_sample){
    if (j == 1){
      T[j] <-  10^as.numeric(solution[7])}
    else {
      T[j] <- T[j-1]*as.numeric(solution[6+j])
    }
  }

  r04i <- r08i/r48i
  U8i <- U_init*1e-6/238*6.02e23
  U4i <- r48i*l238/l234*U8i
  Th0i <- r04i*l234/l230*U4i

  F238 <- f238*l238*U8i
  F234 <- f234*l234*U4i
  F230 <- f230*l230*Th0i

  a4 <- l234 + k234
  a8 <- l238 + k238
  a0 <- l230 + k230
  A <- l230*l234/((a0 - a8)*(a4 - a8)) - l230*l234*F238/(l238*U8i*a8*(a4 - a8)*(a0 - a8))
  B <- l230/(a0 - a4)*r48i - l230*l234/((a0 - a4)*(a4 - a8)) - l230*F234/(l238*U8i*a4*(a0 - a4)) - l230*l234*F238/(l238*U8i*a4*(a0 - a4)*(a4 - a8))
  C <- r08i - l230/(a0 - a4)*r48i + l230*l234/((a0 - a4)*(a4 - a8)) - l230*l234/((a0 - a8)*(a4 - a8)) - F230/(l238*U8i*a0) - l230*F234/(l238*U8i*a0*a4) + l230*F234/(l238*U8i*a4*(a0 - a4)) - l230*l234*F238/(l238*U8i*a0*a4*a8) + l230*l234*F238/(l238*U8i*a4*a8*(a0 - a4)) - l230*l234*F238/(l238*U8i*a8*(a0 - a4)*(a4 - a8)) + l230*l234*F238/(l238*U8i*a8*(a0 - a8)*(a4 - a8))

  R48_calc <- vector(mode="numeric", length=nb_sample)
  R08_calc <- vector(mode="numeric", length=nb_sample)

  for (j in 1:nb_sample){
    R48_calc[j] <- ((l234/(a4 - a8) - l234*F238/(l238*U8i*a8*(a4 - a8)))*exp(-a8*T[j]) + (r48i - l234/(a4 - a8) - l234*F238/(l238*U8i*a4*(a4 - a8)) - F234/(l238*U8i*a4))*exp(-a4*T[j]) + F234/(l238*U8i*a4) + l234*F238/(l238*U8i*a4*a8))/((1 - F238/(l238*U8i*a8))*exp(-a8*T[j]) + F238/(l238*U8i*a8))
    R08_calc[j] <- (A*exp(-a8*T[j]) + B*exp(-a4*T[j]) + C*exp(-a0*T[j]) + l230*l234*F238/(l238*U8i*a0*a4*a8) + l230*F234/(l238*U8i*a0*a4) + F230/(l238*U8i*a0))/((1 - F238/(l238*U8i*a8))*exp(-a8*T[j]) + F238/(l238*U8i*a8))
  }

  data <- as.data.frame(cbind(R48_calc, R08_calc, T))
  data <- cbind(data, df$Depth..m.[1:nb_sample])
  colnames(data) <- c("R234U_238U", "R230Th_238U", "T", "Depth")


  # graphs ------------------------------------------------------------------

  theme_plots <- ggplot2::theme(plot.title = ggplot2::element_text(face = "bold", size = (15), hjust = 0.5),
                       legend.title=ggplot2::element_blank(),
                       axis.title.y = ggplot2::element_text(size=18),
                       axis.title.x = ggplot2::element_text(size=18),
                       legend.text=ggplot2::element_text(size=12)) +
    ggplot2::theme_bw() +
    ggplot2::theme(panel.grid.major = ggplot2::element_blank(), panel.grid.minor = ggplot2::element_blank())


  # (234U/238U) vs depth
  p48 <- ggplot2::ggplot(target_ratios, ggplot2::aes(R48_target,depth)) +
    ggplot2::geom_errorbarh(ggplot2::aes(xmax = R48_target+R48_err, xmin = R48_target-R48_err, height=0.02)) + # plot error bars
    ggplot2::geom_point(size = 5, colour = 'blue') +
    ggplot2:: geom_point(data=data, ggplot2::aes(R234U_238U,Depth), size = 5, colour = 'red') + ggplot2::scale_y_reverse() +
    theme_plots +  ggplot2::xlab(expression("("^234*"U/"^238*"U)")) +
    ggplot2::ylab('Depth (m)')

  # (230Th/238U) vs depth
  p08 <- ggplot2::ggplot(target_ratios, ggplot2::aes(R08_target,depth)) +
    ggplot2::geom_errorbarh(ggplot2::aes(xmax = R08_target+R08_err, xmin = R08_target-R08_err, height=0.02)) + # plot error bars
    ggplot2::geom_point(size = 5, colour = 'blue') +
    ggplot2::geom_point(data=data, ggplot2::aes(R230Th_238U,Depth), size = 5, colour = 'red') + ggplot2::scale_y_reverse() +
    theme_plots +  ggplot2::xlab(expression("("^230*"Th/"^238*"U)")) +
    ggplot2::ylab('Depth (m)')

  # Weathering age vs depth
  pt <- ggplot2::ggplot(data, ggplot2::aes(T/1000,Depth)) +
    ggplot2::geom_point(size = 5, colour = 'red') + ggplot2::scale_y_reverse() + theme_plots +
    ggplot2::xlab('Weathering age (ka)') +
    ggplot2::ylab('Depth (m)')

  # if(!(require(cowplot))){
  #   install.packages("cowplot")
  #   library(cowplot)
  # }

  print_plot <- cowplot::plot_grid(p48, p08, pt, ncol = 3)

  # finish ------------------------------------------------------------------
  # Regolith production rate (mm/kyr)
  rate <- (depth[nb_sample]-depth[1])*1e6/(T[1] - T[nb_sample])
  print(print_plot)
  solution_df <- as.data.frame(t(c(k238, k234/k238, f238/k238, f234/f238, r48i, r08i)))
  solution_list <- list(regolith_production_rate = rate,
                        parameters = solution_df,
                        calculated_ratios_and_ages = data,
                        graphs = print_plot, nb_Sample = nb_sample,
                        min_val = min_val, max_val = max_val)

  return(solution_list)
}



U_series_gain_loss_optimisation/app.R
require(waiter)
require(DT)
require(ggplot2)
require(cowplot)
require(GA)

library(shiny)

waiting_screen <- tagList(
    spin_solar(),
    HTML("<br><br><br><br><br>Generating model output...")
)

ui <- bootstrapPage(
    mainPanel(
        # ui <- fluidPage(
        use_waiter(),
        # Application title
        titlePanel("U-series isotopes gain loss model - genetic algorithm optimisation"),

        tabsetPanel(id = "abset",
                    # Load data panel
                    tabPanel("Load the data",
                             p("Before uploading, check that your CSV file contains columns with these names:"),
                             HTML("
                            <li> <b>Sample ID</b>: sample identification code
                            <li> <b>Depth (m)</b>: sample depth (in m)
                            <li> <b>RU234_U238</b>: [<sup>234</sup>U/<sup>238</sup>U] activity ratios
                            <li> <b>R48_2SE</b>: the 2 sigma errors of the [<sup>234</sup>U/<sup>238</sup>U] activity ratios
                            <li> <b>RTh230_U238</b>: [<sup>230</sup>Th/<sup>238</sup>U] activity ratios
                            <li> <b>R08_2SE</b>: the 2 sigma errors of the [<sup>230</sup>Th/<sup>238</sup>U] activity ratios <br><br>
                            The deepest sample is taken at the starting composition, so make sure that what you want to use as starting material is included and assigned the deepest depth <br><br>
                            Please cite:<br> <b>Dosseto, A., et al. (2014). 'Age and weathering rate of sediments in small catchments: The role of hillslope erosion.' Geochimica et Cosmochimica Acta 132(0): 238-258 </b>
                            "),
                             tags$hr(),
                             fileInput("file1",
                                       "Choose CSV file",
                                       accept = c("text/csv",
                                                  "text/comma-separated-values,text/plain",
                                                  ".csv")),
                             actionButton('gotoinspect', 'Go to inspect the data')
                    ), # end of tab

                    # Inspect data panel
                    tabPanel("Inspect the data",
                             value = "inspectthedata",
                             p("Here is the raw data from the CSV file"),
                             DT::dataTableOutput('contents'),
                             actionButton('gotosetmodel', 'Go to set the model parameters')
                    ), # end of tab

                    # Set model parameters panel
                    tabPanel("Set model parameters",
                             value = "setmodelparameters",
                             fluidRow(
                                 column(4,
                                        numericInput("nbit", "Number of iterations:", 1000, min = 1, max = 1e6),
                                        numericInput("logT_min", "Minimum log10 of weathering age of top sample (in log(yr)):", 2, min = 1, max = 7),
                                        numericInput("logT_max", "Maximum log10 of weathering age of top sample (in log(yr)):", 6, min = 1, max = 7),
                                        numericInput("Tmult_min", "Minimum fraction weathering age of sample compared to that of sample directly above:", 0.1, min = 0.001, max = 1),
                                 ),
                                 column(4,
                                        numericInput("logk238_min", "Minimum log10 of k238 (in log(yr-1)):", -8, min = -20, max = 0),
                                        numericInput("logk238_max", "Maximum log10 of k238 (in log(yr-1)):", -3, min = -20, max = 0),
                                        numericInput("k48_min", "Minimum k234/k238:", 0.1, min = 0.01, max = 1000),
                                        numericInput("k48_max", "Maximum k234/k238:", 10, min = 0.01, max = 1000)
                                 ),
                                 column(4,
                                        numericInput("logf238_k238_min", "Minimum log10 of f238/k238:", -8, min = -20, max = 3),
                                        numericInput("logf238_k238_max", "Maximum log10 of f238/k238:", 3, min = -20, max = 3),
                                        numericInput("f48_min", "Minimum f234/f238:", 0.1, min = 0.1, max = 1000),
                                        numericInput("f48_max", "Maximum f234/f238:", 10, min = 0.1, max = 1000),
                                        # save solutions
                                        textInput(inputId = "filename", label = "File name to save calculated ratios and weathering ages to", value = "profile_name")
                                 )),
                             actionButton("run", label = "Run simulation and visualise the output")
                    ),

                    # Visualise results tab
                    tabPanel("Visualise the model",
                             value = "visualise",
                             tags$hr(),
                             textOutput("print_results"),
                             tags$hr(),
                             textOutput("param_sol"),
                             tableOutput("print_param"),
                             # show a spinner while we wait for the plots to draw
                             #  withSpinner()
                             # # Input: Choose dataset ----
                             # selectInput("dataset", "Choose a dataset:",
                             #             choices = c("rock", "pressure", "cars")),
                             #
                             # Button
                             downloadButton("downloadData", "Download"),
                             tags$hr(),
                             plotOutput("plots",
                                        width = "100%",
                                        height = "300px"),
                             color="blue",
                             size = 5
                    ) # end of tab

        )
    ))

# Define server logic
server <- function(input, output, session) {
    # activate the buttons to move between tabs
    observeEvent(input$gotoinspect, {
        updateTabsetPanel(session, "abset",
                          selected = "inspectthedata")
    })

    output$contents <- DT::renderDataTable({
        inFile <- input$file1

        if (is.null(inFile))
            return(DF)

        read.csv(inFile$datapath)
    })

    observeEvent(input$gotosetmodel, {
        updateTabsetPanel(session, "abset",
                          selected = "setmodelparameters"  )
    })

    observeEvent(input$run, {
        updateTabsetPanel(session, "abset",
                          selected = "visualise"  )
    })


    observeEvent(input$run, {
        waiter_show(
            html = waiting_screen,
            color = "#333e48",
            id = 'plots',
            hide_on_render = TRUE
        )

        showNotification("Starting model run...")

        # output_model <-  reactive({
        inFile <- input$file1
        if (is.null(inFile)) {
            input_data <- DF
        } else {
            input_data <- read.csv(inFile$datapath)
        }

        output_model <- UseriesGAwigl::U_series_gain_loss_function(input_data, nbit = input$nbit,
                                                                   logT_min = input$logT_min, logT_max = input$logT_max,
                                                                   Tmult_min = input$Tmult_min,
                                                                   logk238_min = input$logk238_min, logk238_max = input$logk238_max,
                                                                   k48_min = input$k48_min, k48_max = input$k48_max,
                                                                   logf238_k238_min = input$logf238_k238_min, logf238_k238_max = input$logf238_k238_max,
                                                                   f48_min = input$f48_min, f48_max = input$f48_max)
        # output_model
        # })

        showNotification("Model run complete.")


        output$print_results <- renderText({ paste("Regolith production rate (mm/kyr):",
                                                   round(output_model$regolith_production_rate,1)) })

        output$param_sol <- renderText("Solutions:")

        output_model$parameters[c(1,3)] <- log10(output_model$parameters[c(1,3)])


        names(output_model$parameters) <- c("log10(k238)", "k234/k238", "log10(f238/k238)",
                                            "f234/f238", "r48i", "r08i")

        output$print_param <- renderTable({ output_model$parameters },
                                          colnames = TRUE)

        names(output_model$calculated_ratios_and_ages) <- c("[234U/238U]", "[230Th/238U]",
                                                            "Weathering age (yr)", "Depth (m)")

        # save solutions for weathering ages
        output$downloadData <- downloadHandler(
            filename = function() {
                paste(input$filename, ".csv", sep = "")
            },
            content = function(file) {
                write.csv(output_model$calculated_ratios_and_ages, file, row.names = F)
            }
        )

        # # draw the plots
        output$plots <- renderPlot({
            # print(output_model$graph_t) })
            #     # draw plots in a panel
            print(output_model$graphs)
        })
        # })
    })
}

# Run the application
shinyApp(ui = ui, server = server)
